velox/vector/arrow/Bridge.cpp

```diff
@@ -315,68 +227,328 @@ const char* exportArrowFormatStr(const TypePtr& type) {
+    vector_size_t j = 0, k = 0;
+    rows.apply([&](vector_size_t i) {
+      rawOffsets[j++] = k;
+      if (!vec.isNullAt(i)) {
+        childRows.addRange(vec.offsetAt(i), vec.sizeAt(i));
+        k += vec.sizeAt(i);
+      }
+    });
+    VELOX_DCHECK_EQ(j, out.length);
+    rawOffsets[j] = k;
+  }
+  holder.setBuffer(1, offsets);
+  out.n_buffers = 2;
+}
+
+void exportArrays(
+    const ArrayVector& vec,
+    const Selection& rows,
+    ArrowArray& out,
+    memory::MemoryPool* pool,
+    VeloxToArrowBridgeHolder& holder) {
+  Selection childRows(vec.elements()->size());
+  exportOffsets(vec, rows, out, pool, holder, childRows);
+  holder.resizeChildren(1);
+  exportBase(
+      *vec.elements()->loadedVector(),
+      childRows,
+      *holder.allocateChild(0),
+      pool);
+  out.n_children = 1;
+  out.children = holder.getChildrenArrays();
+}
+
+void exportMaps(
+    const MapVector& vec,
+    const Selection& rows,
+    ArrowArray& out,
+    memory::MemoryPool* pool,
+    VeloxToArrowBridgeHolder& holder) {
+  RowVector child(
+      pool,
+      ROW({"key", "value"}, {vec.mapKeys()->type(), vec.mapValues()->type()}),
+      nullptr,
+      vec.mapKeys()->size(),
+      {vec.mapKeys(), vec.mapValues()});
+  Selection childRows(child.size());
+  exportOffsets(vec, rows, out, pool, holder, childRows);
+  holder.resizeChildren(1);
+  exportBase(child, childRows, *holder.allocateChild(0), pool);
+  out.n_children = 1;
+  out.children = holder.getChildrenArrays();
+}
+
+void exportBase(
+    const BaseVector& vec,
+    const Selection& rows,
+    ArrowArray& out,
+    memory::MemoryPool* pool) {
+  auto holder = std::make_unique<VeloxToArrowBridgeHolder>();
+  out.buffers = holder->getArrowBuffers();
+  out.length = rows.count();
+  out.offset = 0;
+  out.dictionary = nullptr;
+  exportNulls(vec, rows, out, pool, *holder);
+  switch (vec.encoding()) {
+    case VectorEncoding::Simple::FLAT:
+      exportFlat(vec, rows, out, pool, *holder);
+      break;
+    case VectorEncoding::Simple::ROW:
+      exportRows(*vec.asUnchecked<RowVector>(), rows, out, pool, *holder);
       break;
+    case VectorEncoding::Simple::ARRAY:
+      exportArrays(*vec.asUnchecked<ArrayVector>(), rows, out, pool, *holder);
+      break;
+    case VectorEncoding::Simple::MAP:
+      exportMaps(*vec.asUnchecked<MapVector>(), rows, out, pool, *holder);
+      break;
+    default:
+      VELOX_NYI("{} cannot be exported to Arrow yet.", vec.encoding());
   }
+  out.private_data = holder.release();
+  out.release = bridgeRelease;
+}
 
-  // TODO: No dictionaries for now.
-  arrowArray.dictionary = nullptr;
+} // namespace
 
-  // We release the unique_ptr since bridgeHolder will now be carried inside
-  // ArrowArray.
-  arrowArray.private_data = bridgeHolder.release();
+void exportToArrow(
+    const VectorPtr& vector,
+    ArrowArray& arrowArray,
+    memory::MemoryPool* pool) {
+  exportBase(*vector, Selection(vector->size()), arrowArray, pool);
```

Comment:
we do we need this Selection? It would be nice to explain it in a high-level in the PR summary.

Reply:
Sure I will add the explanation.